# 📘 Port Authority Capstone
## Phase 1 – Data Audit & Structural Engineering

This notebook performs structural validation of all raw datasets.

The purpose of this structural audit is to validate the integrity, granularity, and analytical readiness of all Port Authority datasets prior to any descriptive or exploratory analysis. 

This phase ensures:
- Consistent data types
- Structural completeness
- Reliable temporal alignment
- Dataset-level grain clarity
- Preservation of original analytical characteristics

No analytical transformations were performed during this stage.

In [1]:
# =========================================
# Project Setup
# =========================================

import pandas as pd
import numpy as np
import os

pd.set_option("display.max_columns", None)

print("Environment Ready")


Environment Ready


In [2]:
# =========================================
# Load Raw Datasets
# =========================================

traffic_df = pd.read_csv("../data/raw/All Recorded Traffic.txt", sep="\t")
bus_df = pd.read_csv("../data/raw/All Recorded PABT Bus.txt", sep="\t")
passenger_df = pd.read_csv("../data/raw/All Recorded PABT Passenger.txt", sep="\t")
speeds_df = pd.read_csv("../data/raw/Facility Mobility Speeds.txt", sep="\t")

print("Datasets Loaded Successfully")


Datasets Loaded Successfully


# 🚦 Traffic Dataset – Structural Audit

In [11]:
# Shape
print("Traffic Shape:", traffic_df.shape)

# Data types
print("\nTraffic Data Types:")
print(traffic_df.dtypes)

# Duplicate rows
print("Traffic Duplicate Rows:", traffic_df.duplicated().sum())

# Missing values
traffic_missing = traffic_df.isna().sum()
print("\nTraffic Missing Values:")
print(traffic_missing[traffic_missing > 0])

# DATE preview
print("\nTraffic DATE Sample:")
print(traffic_df["DATE"].head())


Traffic Shape: (5383378, 30)

Traffic Data Types:
DAY                   int64
DATE         datetime64[ns]
FAC                   int64
LANE                  int64
TIME                  int64
TOTAL                 int64
CLASS 1             float64
CLASS 2             float64
CLASS 3             float64
CLASS 4             float64
CLASS 5             float64
CLASS 6             float64
CLASS 7             float64
CLASS 8             float64
CLASS 9             float64
CLASS 11            float64
CASH                  int64
EZPASS                int64
VIOLATION             int64
LANEMODE             object
Month                 int64
FAC_B                object
Autos                 int64
Small_T               int64
Large_T               int64
Buses                 int64
Yr                    int64
FAC_G                object
FAC_G2               object
Day_Name             object
dtype: object
Traffic Duplicate Rows: 0

Traffic Missing Values:
CLASS 1       12710
CLASS 2      407007
CLASS

### Traffic Dataset – Structural Findings

The Traffic dataset contains 5,383,378 rows and 30 columns with no duplicate records.

Each row represents traffic counts for a specific facility (FAC), lane, date, and time interval. This indicates a highly granular operational dataset at the lane level.

During inspection, the following observations were made:

- Data types were consistent with expected structure.
- All operational traffic flow variables (TOTAL, CASH, EZPASS, Autos, etc.) are structurally complete with zero missing values.
- Date format required conversion to proper datetime format for temporal alignment.
- Missing values (NULL) are concentrated exclusively in the detailed vehicle classification columns (CLASS 1–CLASS 11), with the highest missingness observed in CLASS 6 (1,354,198 records). This indicates inconsistent capture of granular vehicle segmentation while overall traffic totals remain reliable.
- The NULL values were preserved intentionally, as they represent operational gaps rather than structural corruption.
- Column naming conventions include abbreviated identifiers (e.g., FAC, FAC_G2). These were preserved to maintain raw dataset integrity and traceability.

From a structural standpoint, this dataset operates at a highly granular lane-level resolution. 
Any future integration or aggregation will require careful consideration of its temporal and facility-level structure.

In [ ]:
# 🚌 Bus Dataset – Structural Audit

In [4]:
# Shape
print("Bus Shape:", bus_df.shape)

# Data types
print("\nBus Data Types:")
print(bus_df.dtypes)

# Duplicate rows
print("\nBus Duplicate Rows:", bus_df.duplicated().sum())

# Missing values
print("\nBus Missing Values:")
print(bus_df.isna().sum())

# Bus DATE preview
print("\nBus Start_Date Sample:")
print(bus_df["Start_Date"].head())

print("\nBus End_Date Sample:")
print(bus_df["End_Date"].head())

Bus Shape: (2784, 4)

Bus Data Types:
Start_Date     object
End_Date       object
Carrier        object
Volume        float64
dtype: object

Bus Duplicate Rows: 0

Bus Missing Values:
Start_Date    0
End_Date      0
Carrier       0
Volume        0
dtype: int64

Bus Start_Date Sample:
0    2021-02-08 00:00:00.000
1    2021-02-08 00:00:00.000
2    2021-02-08 00:00:00.000
3    2021-02-08 00:00:00.000
4    2021-02-08 00:00:00.000
Name: Start_Date, dtype: object

Bus End_Date Sample:
0    2021-02-12 00:00:00.000
1    2021-02-12 00:00:00.000
2    2021-02-12 00:00:00.000
3    2021-02-12 00:00:00.000
4    2021-02-12 00:00:00.000
Name: End_Date, dtype: object


# 👥 PASSENGER DATASET AUDIT

In [5]:
# Shape
print("Passenger Shape:", passenger_df.shape)

# Data types
print("\nPassenger Data Types:")
print(passenger_df.dtypes)

# Duplicate rows
print("\nPassenger Duplicate Rows:", passenger_df.duplicated().sum())

# Missing values
print("\nPassenger Missing Values:")
print(passenger_df.isna().sum())

# Passenger DATE preview
print("\nPassenger Start_Date Sample:")
print(passenger_df["Start_Date"].head())

print("\nPassenger End_Date Sample:")
print(passenger_df["End_Date"].head())

Passenger Shape: (2786, 4)

Passenger Data Types:
Start_Date     object
End_Date       object
Carrier        object
Volume        float64
dtype: object

Passenger Duplicate Rows: 1

Passenger Missing Values:
Start_Date    2
End_Date      2
Carrier       2
Volume        2
dtype: int64

Passenger Start_Date Sample:
0    2020-12-07 00:00:00.000
1    2020-12-07 00:00:00.000
2    2020-12-07 00:00:00.000
3    2020-12-07 00:00:00.000
4    2020-12-07 00:00:00.000
Name: Start_Date, dtype: object

Passenger End_Date Sample:
0    2020-12-11 00:00:00.000
1    2020-12-11 00:00:00.000
2    2020-12-11 00:00:00.000
3    2020-12-11 00:00:00.000
4    2020-12-11 00:00:00.000
Name: End_Date, dtype: object


# 🚗 SPEEDS DATASET AUDIT

In [6]:
# Shape
print("Speeds Shape:", speeds_df.shape)

# Data types
print("\nSpeeds Data Types:")
print(speeds_df.dtypes)

# Duplicate rows
print("\nSpeeds Duplicate Rows:", speeds_df.duplicated().sum())

# Missing values
print("\nSpeeds Missing Values:")
print(speeds_df.isna().sum())

# Speeds Month_Year preview
print("\nSpeeds Month_Year Sample:")
print(speeds_df["Month_Year"].head())

Speeds Shape: (192, 7)

Speeds Data Types:
Facility           object
Freeflow          float64
Month_Year         object
Avg_Speed         float64
Direction          object
Delta             float64
Facility_Order      int64
dtype: object

Speeds Duplicate Rows: 0

Speeds Missing Values:
Facility          0
Freeflow          0
Month_Year        0
Avg_Speed         0
Direction         0
Delta             0
Facility_Order    0
dtype: int64

Speeds Month_Year Sample:
0    2025-03-01 00:00:00.000
1    2025-03-01 00:00:00.000
2    2025-03-01 00:00:00.000
3    2025-03-01 00:00:00.000
4    2025-03-01 00:00:00.000
Name: Month_Year, dtype: object


### Bus, Passenger & Speeds – Structural Findings

The Bus dataset contains no duplicate records and no missing values. However, the date columns (Start_Date and End_Date) were stored as object type at the time of inspection and therefore required conversion to proper datetime format before analytical processing. Each row represents aggregated bus volume by carrier within a defined date range, indicating a carrier-level aggregation.

The Passenger dataset contains one duplicate record and two structurally incomplete rows identified during inspection. These were flagged for removal prior to descriptive analysis to prevent distortion of summary statistics. Similar to the Bus dataset, its date columns were stored as object type and required conversion to datetime format. The dataset operates at a carrier-level aggregation within defined date ranges.

The Speeds dataset contains no duplicate records and no missing values. The Month_Year column was stored as object type and required conversion to datetime format to ensure proper temporal alignment. Each row represents facility-level mobility metrics at a monthly aggregation.

From a structural perspective, the Bus and Passenger datasets share similar carrier-level aggregation within defined date ranges, making them structurally compatible for future integration after temporal standardization. The Speeds dataset operates at a facility-level monthly aggregation and can be aligned with aggregated Traffic data through facility and time harmonization.
    
No additional structural inconsistencies were identified at this stage. Necessary structural corrections are applied in the following section prior to descriptive and exploratory analysis.

## 🔧 STRUCTURAL CORRECTIONS

In [7]:
# --- Traffic ---
traffic_df["DATE"] = pd.to_datetime(traffic_df["DATE"])

# --- Bus ---
bus_df["Start_Date"] = pd.to_datetime(bus_df["Start_Date"])
bus_df["End_Date"] = pd.to_datetime(bus_df["End_Date"])

# --- Passenger ---
passenger_df["Start_Date"] = pd.to_datetime(passenger_df["Start_Date"])
passenger_df["End_Date"] = pd.to_datetime(passenger_df["End_Date"])

# Remove Passenger duplicate
passenger_df = passenger_df.drop_duplicates()

# Remove Passenger incomplete rows
passenger_df = passenger_df.dropna()

# --- Speeds ---
speeds_df["Month_Year"] = pd.to_datetime(speeds_df["Month_Year"])

print("All structural corrections applied successfully.")

All structural corrections applied successfully.


In [8]:
# ===============================
# Structural Verification Check
# ===============================

print("Traffic DATE dtype:", traffic_df["DATE"].dtype)
print("Bus Start_Date dtype:", bus_df["Start_Date"].dtype)
print("Bus End_Date dtype:", bus_df["End_Date"].dtype)
print("Passenger Start_Date dtype:", passenger_df["Start_Date"].dtype)
print("Passenger End_Date dtype:", passenger_df["End_Date"].dtype)
print("Speeds Month_Year dtype:", speeds_df["Month_Year"].dtype)

print("\nPassenger Duplicate Rows After Cleaning:", passenger_df.duplicated().sum())
print("Passenger Missing Values After Cleaning:")
print(passenger_df.isna().sum())

# =========================================
# Save Structurally Cleaned Datasets
# =========================================

traffic_df.to_csv("../data/processed/traffic_cleaned.csv", index=False)
bus_df.to_csv("../data/processed/bus_cleaned.csv", index=False)
passenger_df.to_csv("../data/processed/passenger_cleaned.csv", index=False)
speeds_df.to_csv("../data/processed/speeds_cleaned.csv", index=False)

print("Cleaned datasets saved successfully to processed folder.")

Traffic DATE dtype: datetime64[ns]
Bus Start_Date dtype: datetime64[ns]
Bus End_Date dtype: datetime64[ns]
Passenger Start_Date dtype: datetime64[ns]
Passenger End_Date dtype: datetime64[ns]
Speeds Month_Year dtype: datetime64[ns]

Passenger Duplicate Rows After Cleaning: 0
Passenger Missing Values After Cleaning:
Start_Date    0
End_Date      0
Carrier       0
Volume        0
dtype: int64
Cleaned datasets saved successfully to processed folder.


## Structural Corrections Completed

The following structural changes were applied before starting descriptive analysis and EDA:

- Converted all date-related columns to proper datetime format:
  - Traffic: DATE
  - Bus: Start_Date, End_Date
  - Passenger: Start_Date, End_Date
  - Speeds: Month_Year

This ensures accurate time-based analysis, filtering, and aggregation.

- Removed one duplicate record from the Passenger dataset.
- Removed two incomplete rows from the Passenger dataset.

These adjustments were made because duplicate and incomplete records can distort summary statistics and analytical results.

The missing values in the Traffic dataset (vehicle CLASS columns) were intentionally not treated. These represent analytical missingness in detailed vehicle classification and are part of the original data structure. Modifying them at this stage could alter the analytical characteristics of the dataset.

    At this point, only structural corrections have been performed. No analytical patterns have been modified.The structurally corrected datasets have been saved in the processed directory and will be used for all subsequent descriptive and exploratory analysis. The raw datasets remain unchanged.
---